src/providers/hana/qgshanaconnection.cpp

```diff
@@ -0,0 +1,791 @@
+        srid = -1;
+        break;
+      }
+      prevSrid = srid;
+    }
+
+    rsSrid->close();
+  }
+  catch ( const Exception &ex )
+  {
+    throw QgsHanaException( ex.what() );
+  }
+
+  return srid;
+}
+
+QStringList QgsHanaConnection::getLayerPrimaryeKeys( const QgsHanaLayerProperty &layerProperty )
+{
+  QStringList ret;
+
+  try
+  {
+    DatabaseMetaDataUnicodeRef dbmd = mConnection->getDatabaseMetaDataUnicode();
+    ResultSetRef rsPrimaryKeys = dbmd->getPrimaryKeys( nullptr,
+                                 QgsHanaUtils::toUtf16( layerProperty.schemaName ),
+                                 QgsHanaUtils::toUtf16( layerProperty.tableName ) );
+
+    while ( rsPrimaryKeys->next() )
+    {
+      auto keyName = rsPrimaryKeys->getNString( 4 );
+      QString clmName = QgsHanaUtils::toQString( keyName );
+      ResultSetRef rsColumns = dbmd->getColumns( nullptr,
+                               QgsHanaUtils::toUtf16( layerProperty.schemaName ),
+                               QgsHanaUtils::toUtf16( layerProperty.tableName ),
+                               QgsHanaUtils::toUtf16( clmName ) );
+
+      if ( rsColumns->next() )
+      {
+        Short dataType = rsColumns->getShort( 5 );
+        short dt = *dataType;
+        if ( dt == SQLDataTypes::TinyInt || dt == SQLDataTypes::SmallInt ||
+             dt == SQLDataTypes::Integer || dt == SQLDataTypes::BigInt )
+          ret << clmName;
+      }
+      rsColumns->close();
+    }
+    rsPrimaryKeys->close();
+  }
+  catch ( const Exception &ex )
+  {
+    throw QgsHanaException( ex.what() );
+  }
+
+  return ret;
+}
+
+QgsWkbTypes::Type QgsHanaConnection::getLayerGeometryType( const QgsHanaLayerProperty &layerProperty )
+{
+  if ( layerProperty.geometryColName.isEmpty() )
+    return QgsWkbTypes::NoGeometry;
+
+  QgsWkbTypes::Type ret;
+
+  QString sql = QStringLiteral( "SELECT upper(%1.ST_GeometryType()), %1.ST_Is3D(), %1.ST_IsMeasured() FROM %2.%3 "
+                                "WHERE %1 IS NOT NULL LIMIT %4" ).arg(
+                  QgsHanaUtils::quotedIdentifier( layerProperty.geometryColName ),
+                  QgsHanaUtils::quotedIdentifier( layerProperty.schemaName ),
+                  QgsHanaUtils::quotedIdentifier( layerProperty.tableName ),
+                  QString::number( GEOM_TYPE_SELECT_LIMIT ) );
+
+  try
+  {
+    StatementRef stmt = mConnection->createStatement();
+    ResultSetRef rsGeomInfo = stmt->executeQuery( QgsHanaUtils::toUtf16( sql ) );
+    QgsWkbTypes::Type geomType = QgsWkbTypes::Unknown, prevGeomType = QgsWkbTypes::Unknown;
+    while ( rsGeomInfo->next() )
+    {
+      geomType = QgsWkbTypes::singleType( QgsHanaUtils::toWkbType(
+                                            rsGeomInfo->getString( 1 ), rsGeomInfo->getInt( 2 ), rsGeomInfo->getInt( 3 ) ) );
+      if ( prevGeomType != QgsWkbTypes::Unknown && geomType != prevGeomType )
+      {
+        geomType = QgsWkbTypes::Unknown;
+        break;
+      }
+      prevGeomType = geomType;
+    }
```

Comment:
Maybe we can simplify that like the SRID detection. But I'll leave that to you.

Reply:
Reworked in 2267454.